core/src/main/scala/org/allenai/pipeline/Pipeline.scala

```diff
@@ -0,0 +1,364 @@
+      } yield pp.stepInfo
+    require(nonExistentDependencies.isEmpty, {
+      val targetNames = targetStepInfo.map(_.className).mkString(",")
+      val dependencyNames = nonExistentDependencies.map(_.className).mkString(",")
+      s"Cannot run steps [$targetNames]. Upstream dependencies [$dependencyNames] have not been computed"
+    })
+    runPipelineReturnResults(title, targets)
+  }
+
+  protected[this] def runPipelineReturnResults(rawTitle: String, targets: Iterable[(String, Producer[_])]): Iterable[(String, Any)] = {
+    // Order the outputs so that the ones with the fewest dependencies are executed first
+    val outputs = targets.toVector.map {
+      case (name, p) => (Workflow.upstreamDependencies(p).size, name, p)
+    }.sortBy(_._1).map { case (count, name, p) => (name, p) }
+    val result: Seq[(String, Any)] = try {
+      val start = System.currentTimeMillis
+      val result = outputs.map { case (name, p) => (name, p.get) }
+      val duration = (System.currentTimeMillis - start) / 1000.0
+      logger.info(f"Ran pipeline in $duration%.3f s")
+      result
+    } catch {
+      case NonFatal(e) =>
+        logger.error("Untrapped exception", e)
+        List()
+    }
+
+    val title = rawTitle.replaceAll("""\s+""", "-")
+    val today = new SimpleDateFormat("yyyy-MM-dd-HH-mm-ss").format(new Date())
+
+    val workflowArtifact = createOutputArtifact[FlatArtifact](s"summary/$title-$today.workflow.json")
+    val workflow = Workflow.forPipeline(targets.toMap)
+    SingletonIo.json[Workflow].write(workflow, workflowArtifact)
+
+    val htmlArtifact = createOutputArtifact[FlatArtifact](s"summary/$title-$today.html")
+    SingletonIo.text[String].write(workflow.renderHtml, htmlArtifact)
+
+    val signatureArtifact = createOutputArtifact[FlatArtifact](s"summary/$title-$today.signatures.json")
+    val signatureFormat = Signature.jsonWriter
+    val signatures = targets.map { case (s, p) => signatureFormat.write(p.stepInfo.signature) }.toList.toJson
+    signatureArtifact.write { writer => writer.write(signatures.prettyPrint) }
+
+    logger.info(s"Summary written to ${toHttpUrl(htmlArtifact.url)}")
+    result
+  }
+
+  // Generate a hash unique to this Producer/Serialization combination
+  protected def hashId[T, A <: Artifact](
+    p: Producer[T],
+    io: Serializer[T, A] with Deserializer[T, A]
+  ) =
+    p.stepInfo.copy(
+      dependencies = p.stepInfo.dependencies + ("io" -> io)
+    ).signature.id
+
+  // Convert S3 URLs to an http: URL viewable in a browser
+  def toHttpUrl(url: URI): URI = {
+    url.getScheme match {
+      case "s3" | "s3n" =>
+        new java.net.URI("http", s"${
+          url.getHost
+        }.s3.amazonaws.com", url.getPath, null)
+      case "file" =>
+        new java.net.URI(null, null, url.getPath, null)
+      case _ => url
+    }
+  }
+
+  def dryRun(outputDir: File, rawTitle: String, targets: Iterable[(String, Producer[_])] = persistedSteps.toList): Unit = {
```

Comment:
DryRun can also take a list of steps and will only output the diagram for the sub-pipeline with those steps.